### ISSUE TITLE
Hot module replacement (HMR) with react-pixi

### ISSUE THREAD

**Issue by smith-chris** [inlet/react-pixi#145]
Hi all 😄, 

thanks for this wonderful library! I've been using it on my pet projects (shameful pulg ;) : https://www.reddit.com/r/Gameboy/comments/d8bnnf/i_wonder_how_original_pokemon_would_look_like_on/) and its been pleasure to work with

BUT! I'd love to increase my productivity with HMR, and so I tried and I did a little progress with it today until I got to the point where mysterious knowledge of Elders of Webpack&React is needed. Before I'll spend a week on it without any outcome, and decide that I should not be a programmer, I have to ask a community! 
Has anyone had any success with HMR?

The video here explains it best up until which point I got it working:
https://youtu.be/RDdJ2nWtYl0

Here is a link to the repo at exact commit I'm talking about (in case I change it all later): https://github.com/smith-chris/react-pixi-starter/tree/6ff51e2ef8120d494ed4af95db26787481e3b8f5

I hope the repo is self explanatory and I'd be grateful for even any ideas how to approach this problem. 
To me it seems like there has to be a way 'store' components state before replacement and 'restore' it with new components/modules. If it works for react-dom it should be doable for react-pixi. 

I will try to solve this on my own but any help would be appreciated. Thanks!

**Comment by inlet** [comment]
Hi @smith-chris,

Thanks for contributing!
HMR should work, not sure why it isn't working on your side though.

Example with `create-react-app` (module.hot enabled in `index.js`):
![ScreenFlow](https://user-images.githubusercontent.com/232559/65960765-794db680-e455-11e9-8cf8-7aaf9c4c501e.gif)

<details>
<summary>Rectangle.js</summary>

```jsx
import { PixiComponent } from '@inlet/react-pixi';
import { Graphics } from 'pixi.js';

const Rectangle = PixiComponent('Rect', {
  create({x, y, width, height}) {
    const g= new Graphics();
    g.beginFill(0xff0000);
    g.drawRect(x, y, width, height);
    return g;
  },
  applyProps(ins, _, { x, y, width, height }) {
    ins.x = x;
    ins.y = y;
    ins.width = width;
    ins.height = height;
  },
  didMount() {
    console.log('rect mounted')
  },
  willUnmount() {
    console.log(`rect unmounted`)
  }
});
```

</details>

As you can see it unmounts and mount back updating the component state with new props.

**Comment by smith-chris** [comment]
Thanks for reply @inlet!

One thing is that example you gave doesn't have any in-comoponent state and losing component's state is exactly what I'm missing in my hrm solution (you can see that I'm changing prop and it doesn't reload as well here: http://www.youtube.com/watch?v=RDdJ2nWtYl0&t=0m48s but other 3 bunnies lose their state :/)

Could you please provide example repo or steps you've taken to configure CRA with react-pixi & hmr? Or did you just use <Stage/> component that can be embedded within regular react?

Cheers

**Comment by inlet** [comment]
Hi @smith-chris, 

Great to hear it works now. Yes you need to wrap it in `<Stage />` as this is the injection point for the reconciliation.

**Note**: You don't have to use the `<Stage />`, if you'd like you can create your own renderer based on `PixiFiber` which is exported as well.

Best, Patrick

**Comment by smith-chris** [comment]
I think I was too excited about HMR and didn't investigate further to find out that only component markup in `App.js` are properly hot replacing. After digging it for a while I figured that importing and using rhl `hot()` function (in any way actually) enables any component to have proper HMR. I've made a video demonstrating what I mean: https://youtu.be/Yp90SHA5GPs

As you can see React's `AutoCounter`s state does not reset on HMR, but `PixiCounter`s state does reset unless we import `hot()` and use it in any way actuall. 
What do you think about it @inlet ? 
It doesn't feel right to me

**Comment by inlet** [comment]
I see, well `react-dom` is patched to support HMR, see https://github.com/gaearon/react-hot-loader#hot-loaderreact-dom and we're not using `react-dom`'s reconciliation which explains why the Pixi component isn't hot reloaded by default.

If you know how we can implement auto-HMR please let me know (or create a PR), but I think this requires some changes on the reconciliation side (only in development mode). But honestly, I don't have an idea how we could make this work.

For now best is to hot export your Pixi components which you'd like to support HMR.

**Comment by smith-chris** [comment]
TL;DR
- auto-hmr with <Stage/> example: https://github.com/smith-chris/react-pixi-starter/tree/00d4e322a361ecd9bb289728073d5db57b3794ef
- auto-hmr with react-pixi render() function example: https://github.com/smith-chris/react-pixi-starter/tree/hrm-no-stage
- CRA auto-hmr: couldn't get it to work
- what changes were needed for the repo from original post (for curious ones ;)): https://github.com/smith-chris/react-pixi-starter/commit/6f1c9d2911780ff7d7628757bff910beaeb197b3
- I'll keep you update if I'll find more errors :)
 
Full update:

After a lot of trial & error I achieved full auto-hmr with <Stage/> component :) here's where I have it working (also with Redux configured correctly as its config is not straightforward): https://github.com/smith-chris/react-pixi-starter/tree/00d4e322a361ecd9bb289728073d5db57b3794ef

While doing that, I discovered that main difference between my starter & CRA that influenced HMR behaviour, it was the inclusion of this part of my webpack config: `alias: { 'react-dom': '@hot-loader/react-dom'  }`

And so I made a minimal changes needed to get HMR properly working for https://github.com/smith-chris/react-pixi-starter/tree/6ff51e2ef8120d494ed4af95db26787481e3b8f5 that I linked in original post. Here is full hrm without <Stage/> https://github.com/smith-chris/react-pixi-starter/tree/hrm-no-stage (I just hope this helps someone, you don't need to use my starter at all, doing it mostly for myself :p): 

As for CRA I didn't get hmr to work properly, mostly because I'm satisfied with my config (CRA ejected config is way too crazy for my simple needs)

I hope it's not confusing that I'm linking specific commits of the same repo ;) It's just a way to make sure the conversation makes sense after I make more changes to it

I imagine, since I don't patch react-dom there might be some issues, but so far everything (including hooks) seems to work great. I'll keep you updated if something pops up!
